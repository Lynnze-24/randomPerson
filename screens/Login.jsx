import React,{useState} from 'react'
import { View, Text,TouchableOpacity,StyleSheet  } from 'react-native'
import { Button } from 'react-native-paper'
import FormInput from '../components/FormInput'

import cusTheme from '../theme/Theme'
import validate,{validateUser,validatePassword} from '../utils/validate'



const Login = ({login}) => {
     const [username, setUsername] = useState('');
     const [password, setPassword] = useState('');
     const[error,setError] = useState(null);

     const changeUsername = (username)=> setUsername(username)
     const changePassword = (password)=> setPassword(password)

     const keyUpUser = ()=>{
        
        if(!(error===1||error===12))return; // don't run if username is valid
       
        const isValidUser = validateUser(username);
       if(isValidUser && error===1)setError(null)
       if(isValidUser && error===12)setError(2)
     }

     const keyUpPassword = ()=>{
        
        if(!(error===2||error===12))return; // don't run if password is valid
        const isValidPassword = validatePassword(password);
       if(isValidPassword && error===2)setError(null)
       if(isValidPassword && error===12)setError(1)
     }

     const validateLogin = () => {
         let errorVal =validate(username,password);
         if(errorVal){
             setError(errorVal)
         }else{
             login()
         }
     }
     


    return (
        <View testID='loginScreen' style={styles.container}>

            <Text style={styles.header}>Random Person</Text>
            <View>
                <FormInput  hasError={(error===1||error===12)?true:false}
                            testID = "usernameInput"
                            onKeyUp = {keyUpUser}
                            isPassword={false}
                            errorMessage = "Invalid Username"
                            changeValue={changeUsername}
                            value={username} 
                            label="Username" />
                <FormInput   errorMessage = "Invalid Password"
                            testID = "passwordInput" 
                             isPassword={true}
                            onKeyUp = {keyUpPassword}
                             hasError={(error===2||error===12)?true:false}
                             changeValue={changePassword} value={password}
                             label="Password" />
                
                <Button contentStyle={styles.contentButton}
                        testID = "loginButton"
                        mode="contained" 
                        style={styles.button}
                        onPress={validateLogin}>
                            Login
                </Button>
            </View>
        </View>
    )
}

const styles = StyleSheet.create({
    container: {
      flex:1,
      backgroundColor: cusTheme.bg,
      alignItems:'center',
      justifyContent:'center'
   },
   header: {
       color:cusTheme.primary,
       fontSize:25,
       textAlign:'center',
       marginBottom:30
   },
   button:{
       marginTop:20,
       width:150,
       borderRadius:20,
       alignSelf:'center',
       fontWeight:600,
       backgroundColor:cusTheme.primary,
   },
   contentButton:{
       height:40,
       lineHeight:40
   }
   
});

export default Login
