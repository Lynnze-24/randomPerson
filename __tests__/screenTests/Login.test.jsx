import {render, fireEvent } from "@testing-library/react-native";
import Login from "../../screens/Login"
import renderer from 'react-test-renderer';



it('matches snapShot', () => {
    const comp = renderer.create( <Login />).toJSON();
    expect(comp).toMatchSnapshot();
});

it('should render component correctly',()=>{ 
    
    const {getByText,getByTestId,queryByText}=render(
                        <Login />
                        );
    
        getByText('Random Person');
        getByTestId('usernameInput')
        getByTestId('passwordInput')
        getByTestId('loginButton')
        expect(queryByText('Invalid Username')).toBe(null);
        expect(queryByText('Invalid Password')).toBe(null);
})

it('should show user error message if user not valid',()=>{ 
    
    const {getByText,getByTestId,queryByText,debug}=render(
                        <Login />
                        );

    fireEvent.changeText(getByTestId('passwordInput'),'a@')
    fireEvent.press(getByTestId('loginButton'))
    getByText('Invalid Username')
    expect(queryByText('Invalid Password')).toBe(null);
       
})

it('should show user error message if password not valid',()=>{ 
    
    const {getByText,getByTestId,queryByText}=render(
                        <Login />
                        );

    fireEvent.changeText(getByTestId('usernameInput'),'aa')
    fireEvent.press(getByTestId('loginButton'))
    getByText('Invalid Password')
    expect(queryByText('Invalid Username')).toBe(null);
       
})

it('should show both user and password error message if both not valid',()=>{ 
    
    const {getByText,getByTestId,queryByText}=render(
                        <Login />
                        );

    fireEvent.press(getByTestId('loginButton'))
    getByText('Invalid Password')
    getByText('Invalid Username')
})

it('should call login function if both user and password valid',()=>{ 
    const login = jest.fn()
    const {getByText,getByTestId,queryByText}=render(
                        <Login login={login} />
                        );
    fireEvent.changeText(getByTestId('usernameInput'),'aa')
    fireEvent.changeText(getByTestId('passwordInput'),'a@')
    fireEvent.press(getByTestId('loginButton'))
    expect(login).toHaveBeenCalled()
})






