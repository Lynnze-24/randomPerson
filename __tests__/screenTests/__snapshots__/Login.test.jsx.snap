// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapShot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fffeee",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="loginScreen"
>
  <Text
    style={
      Object {
        "color": "#120311",
        "fontSize": 25,
        "marginBottom": 30,
        "textAlign": "center",
      }
    }
  >
    Random Person
  </Text>
  <View>
    <View
      style={
        Object {
          "marginTop": 10,
          "maxWidth": "100%",
          "width": 350,
        }
      }
    >
      <View
        pointerEvents="none"
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 6,
            },
            Object {
              "backgroundColor": "#f6f6f6",
              "borderColor": "#120311",
              "borderRadius": 4,
              "borderWidth": 1,
            },
          ]
        }
        testID="text-input-outline"
      />
      <View>
        <View
          style={
            Array [
              Object {
                "paddingBottom": 0,
              },
              Object {
                "minHeight": 48,
                "paddingTop": 8,
              },
            ]
          }
        >
          <View
            collapsable={false}
            pointerEvents="none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
                "zIndex": 3,
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="none"
              style={
                Object {
                  "backgroundColor": "#f6f6f6",
                  "bottom": 4,
                  "left": 10,
                  "maxHeight": 2,
                  "opacity": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 6,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                  ],
                  "width": 12,
                }
              }
            />
            <Text
              collapsable={false}
              maxFontSizeMultiplier={1.5}
              numberOfLines={1}
              style={
                Object {
                  "backgroundColor": "#f6f6f6",
                  "color": "transparent",
                  "fontFamily": "System",
                  "fontSize": 18,
                  "fontWeight": undefined,
                  "left": 18,
                  "maxWidth": -28,
                  "opacity": 0,
                  "paddingHorizontal": 0,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 29,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "scale": 1,
                    },
                    Object {
                      "scaleY": 1,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
            >
              Username
            </Text>
            <Text
              collapsable={false}
              maxFontSizeMultiplier={1.5}
              numberOfLines={1}
              onLayout={[Function]}
              style={
                Object {
                  "color": "#120311",
                  "fontFamily": "System",
                  "fontSize": 18,
                  "fontWeight": undefined,
                  "left": 0,
                  "opacity": 0,
                  "paddingHorizontal": 14,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 28,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "scale": 1,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
            >
              Username
            </Text>
            <Text
              collapsable={false}
              maxFontSizeMultiplier={1.5}
              numberOfLines={1}
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.54)",
                  "fontFamily": "System",
                  "fontSize": 18,
                  "fontWeight": undefined,
                  "left": 0,
                  "opacity": 0,
                  "paddingHorizontal": 14,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 28,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "scale": 1,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
            >
              Username
            </Text>
          </View>
          <TextInput
            editable={true}
            maxFontSizeMultiplier={1.5}
            multiline={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onKeyUp={[Function]}
            placeholder=""
            placeholderTextColor="rgba(0, 0, 0, 0.54)"
            secureTextEntry={false}
            selectionColor="#120311"
            style={
              Array [
                Object {
                  "flexGrow": 1,
                  "margin": 0,
                  "paddingHorizontal": 14,
                  "zIndex": 1,
                },
                Object {
                  "height": 40,
                },
                Object {
                  "paddingBottom": 0,
                  "paddingTop": 0,
                },
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontSize": 18,
                  "fontWeight": undefined,
                  "textAlign": "left",
                  "textAlignVertical": "center",
                },
                false,
                Array [
                  Object {},
                ],
              ]
            }
            testID="usernameInput"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "marginTop": 10,
          "maxWidth": "100%",
          "width": 350,
        }
      }
    >
      <View
        pointerEvents="none"
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 6,
            },
            Object {
              "backgroundColor": "#f6f6f6",
              "borderColor": "#120311",
              "borderRadius": 4,
              "borderWidth": 1,
            },
          ]
        }
        testID="text-input-outline"
      />
      <View>
        <View
          style={
            Array [
              Object {
                "paddingBottom": 0,
              },
              Object {
                "minHeight": 48,
                "paddingTop": 8,
              },
            ]
          }
        >
          <View
            collapsable={false}
            pointerEvents="none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
                "zIndex": 3,
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="none"
              style={
                Object {
                  "backgroundColor": "#f6f6f6",
                  "bottom": 4,
                  "left": 10,
                  "maxHeight": 2,
                  "opacity": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 6,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                  ],
                  "width": 12,
                }
              }
            />
            <Text
              collapsable={false}
              maxFontSizeMultiplier={1.5}
              numberOfLines={1}
              style={
                Object {
                  "backgroundColor": "#f6f6f6",
                  "color": "transparent",
                  "fontFamily": "System",
                  "fontSize": 18,
                  "fontWeight": undefined,
                  "left": 18,
                  "maxWidth": -28,
                  "opacity": 0,
                  "paddingHorizontal": 0,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 29,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "scale": 1,
                    },
                    Object {
                      "scaleY": 1,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
            >
              Password
            </Text>
            <Text
              collapsable={false}
              maxFontSizeMultiplier={1.5}
              numberOfLines={1}
              onLayout={[Function]}
              style={
                Object {
                  "color": "#120311",
                  "fontFamily": "System",
                  "fontSize": 18,
                  "fontWeight": undefined,
                  "left": 0,
                  "opacity": 0,
                  "paddingHorizontal": 14,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 28,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "scale": 1,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
            >
              Password
            </Text>
            <Text
              collapsable={false}
              maxFontSizeMultiplier={1.5}
              numberOfLines={1}
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.54)",
                  "fontFamily": "System",
                  "fontSize": 18,
                  "fontWeight": undefined,
                  "left": 0,
                  "opacity": 0,
                  "paddingHorizontal": 14,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 28,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "scale": 1,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
            >
              Password
            </Text>
          </View>
          <TextInput
            editable={true}
            maxFontSizeMultiplier={1.5}
            multiline={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onKeyUp={[Function]}
            placeholder=""
            placeholderTextColor="rgba(0, 0, 0, 0.54)"
            secureTextEntry={true}
            selectionColor="#120311"
            style={
              Array [
                Object {
                  "flexGrow": 1,
                  "margin": 0,
                  "paddingHorizontal": 14,
                  "zIndex": 1,
                },
                Object {
                  "height": 40,
                },
                Object {
                  "paddingBottom": 0,
                  "paddingTop": 0,
                },
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontSize": 18,
                  "fontWeight": undefined,
                  "textAlign": "left",
                  "textAlignVertical": "center",
                },
                false,
                Object {
                  "marginRight": 44,
                  "paddingRight": 0,
                },
              ]
            }
            testID="passwordInput"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 24,
                "justifyContent": "center",
                "position": "absolute",
                "width": 24,
              },
              Object {
                "right": 12,
                "top": 15,
              },
            ]
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": false,
              }
            }
            accessible={true}
            focusable={true}
            hitSlop={
              Object {
                "bottom": 6,
                "left": 6,
                "right": 6,
                "top": 6,
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Array [
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                    "margin": 6,
                    "overflow": "hidden",
                  },
                  Object {
                    "borderRadius": 18,
                    "height": 36,
                    "width": 36,
                  },
                  undefined,
                  Object {
                    "margin": 0,
                  },
                ],
              ]
            }
            testID="pwToggle"
          >
            <View>
              <Text />
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      collapsable={false}
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "#120311",
          "borderColor": "transparent",
          "borderRadius": 20,
          "borderStyle": "solid",
          "borderWidth": 0,
          "elevation": 2,
          "fontWeight": 600,
          "marginTop": 20,
          "minWidth": 64,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0.75,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 1.5,
          "width": 150,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 20,
            },
          ]
        }
        testID="loginButton"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              Object {
                "height": 40,
                "lineHeight": 40,
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  undefined,
                  Object {
                    "textTransform": "uppercase",
                  },
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  Object {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  undefined,
                ],
              ]
            }
          >
            Login
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
